File: crates/deepsec-cli/src/commands/pr_comment.rs
use crate::config_loader::Context;
use anyhow::{Result, anyhow};
use clap::Args as ClapArgs;
use deepsec_core::{
    FileRecord, Finding, RevalidationVerdict, Severity, runs::list_runs,
    store::load_all_file_records,
};

#[derive(Debug, ClapArgs)]
pub struct Args {
    #[arg(long)]
    pub project_id: String,
    /// Run id to filter net-new findings by. Defaults to the most recent
    /// `process` run.
    #[arg(long)]
    pub run_id: Option<String>,
    /// Drop revalidated false-positive / fixed findings.
    #[arg(long, default_value_t = true)]
    pub real_only: bool,
    /// Write to this path. Defaults to stdout.
    #[arg(long)]
    pub output: Option<std::path::PathBuf>,
    /// Don't render anything if there are no net-new findings.
    #[arg(long)]
    pub skip_empty: bool,
}

pub fn run(args: Args, ctx: &Context) -> Result<()> {
    let run_id = match args.run_id {
        Some(id) => id,
        None => {
            let runs = list_runs(&ctx.data_root, &args.project_id)?;
            runs.into_iter()
                .find(|r| matches!(r.run_type, deepsec_core::RunType::Process))
                .map(|r| r.run_id)
                .ok_or_else(|| anyhow!("no recent process run found"))?
        }
    };

    let records = load_all_file_records(&ctx.data_root, &args.project_id)?;
    let mut rows: Vec<(String, &FileRecord, &Finding)> = Vec::new();
    for rec in &records {
        for f in &rec.findings {
            if f.produced_by_run_id.as_deref() != Some(run_id.as_str()) {
                continue;
            }
            if args.real_only {
                if let Some(rv) = &f.revalidation {
                    if matches!(
                        rv.verdict,
                        RevalidationVerdict::FalsePositive | RevalidationVerdict::Fixed
                    ) {
                        continue;
                    }
                }
            }
            rows.push((rec.file_path.clone(), rec, f));
        }
    }

    if rows.is_empty() && args.skip_empty {
        return Ok(());
    }

    rows.sort_by(|a, b| b.2.severity.rank().cmp(&a.2.severity.rank()));

    let body = render(&run_id, &rows);
    if let Some(out) = args.output {
        if let Some(parent) = out.parent() {
            fs_err::create_dir_all(parent)?;
        }
        fs_err::write(&out, &body)?;
    } else {
        println!("{body}");
    }
    Ok(())
}

fn render(run_id: &str, rows: &[(String, &FileRecord, &Finding)]) -> String {
    use std::fmt::Write;
    let mut out = String::new();
    if rows.is_empty() {
        writeln!(
            &mut out,
            "## deepsec — no net-new findings for run `{run_id}` :white_check_mark:"
        )
        .ok();
        return out;
    }
    let mut by_sev: std::collections::BTreeMap<Severity, usize> = std::collections::BTreeMap::new();
    for (_, _, f) in rows {
        *by_sev.entry(f.severity).or_default() += 1;
    }
    writeln!(
        &mut out,
        "## deepsec — {} net-new finding{} for run `{}`",
        rows.len(),
        if rows.len() == 1 { "" } else { "s" },
        run_id
    )
    .ok();
    let mut parts: Vec<String> = Vec::new();
    for (sev, n) in by_sev.iter().rev() {
        parts.push(format!("{n} × {}", sev.as_str()));
    }
    writeln!(&mut out, "{}\n", parts.join(" · ")).ok();
    for (path, _rec, f) in rows {
        writeln!(
            &mut out,
            "### `{}` — {} **{}**",
            badge(f.severity),
            f.severity.as_str(),
            f.title
        )
        .ok();
        writeln!(&mut out, "- file: `{path}` (lines {:?})", f.line_numbers).ok();
        writeln!(&mut out, "- slug: `{}`", f.vuln_slug).ok();
        writeln!(&mut out, "- confidence: `{:?}`", f.confidence).ok();
        if let Some(rv) = &f.revalidation {
            writeln!(&mut out, "- revalidation: `{:?}`", rv.verdict).ok();
        }
        writeln!(&mut out, "\n{}\n", f.description).ok();
        writeln!(&mut out, "**Fix:** {}\n", f.recommendation).ok();
    }
    out
}

fn badge(sev: Severity) -> &'static str {
    match sev {
        Severity::Critical => "CRIT",
        Severity::High | Severity::HighBug => "HIGH",
        Severity::Medium => "MED ",
        Severity::Bug => "BUG ",
        Severity::Low => "LOW ",
    }
}

File: internal/cli/commands/pr_comment.go
package commands

import (
	"fmt"
	"sort"
	"strings"

	"github.com/noeljackson/deepsec/internal/cli"
	"github.com/noeljackson/deepsec/internal/core"
	"github.com/spf13/cobra"
)

// NewPrCommentCmd renders markdown for net-new findings from one run.
func NewPrCommentCmd(loader func() (*cli.Context, error)) *cobra.Command {
	var projectID, runID, output string
	var realOnly, skipEmpty bool
	cmd := &cobra.Command{
		Use:   "pr-comment",
		Short: "Render markdown for net-new findings from one process run",
		RunE: func(cmd *cobra.Command, _ []string) error {
			ctx, err := loader()
			if err != nil {
				return err
			}
			if runID == "" {
				runs, err := ctx.DataRoot.ListRuns(projectID)
				if err != nil {
					return err
				}
				for _, r := range runs {
					if r.Type == core.RunTypeProcess {
						runID = r.RunID
						break
					}
				}
				if runID == "" {
					return fmt.Errorf("no recent process run found")
				}
			}

			records, err := ctx.DataRoot.LoadAllFileRecords(projectID)
			if err != nil {
				return err
			}
			var rows []prCommentEntry
			for _, rec := range records {
				for i := range rec.Findings {
					f := &rec.Findings[i]
					if f.ProducedByRunID != runID {
						continue
					}
					if realOnly && f.Revalidation != nil {
						v := f.Revalidation.Verdict
						if v == core.VerdictFalsePositive || v == core.VerdictFixed {
							continue
						}
					}
					rows = append(rows, prCommentEntry{path: rec.FilePath, finding: f})
				}
			}
			if len(rows) == 0 && skipEmpty {
				return nil
			}
			sort.SliceStable(rows, func(i, j int) bool {
				return rows[i].finding.Severity.Rank() > rows[j].finding.Severity.Rank()
			})

			body := renderPRComment(runID, rows)
			if output == "" {
				fmt.Println(body)
				return nil
			}
			return writePrivateFile(output, []byte(body))
		},
	}
	cmd.Flags().StringVar(&projectID, "project-id", "", "Project id (required)")
	_ = cmd.MarkFlagRequired("project-id")
	cmd.Flags().StringVar(&runID, "run-id", "", "Run to filter by (default: most recent process run)")
	cmd.Flags().BoolVar(&realOnly, "real-only", true, "Drop FP/Fixed findings")
	cmd.Flags().BoolVar(&skipEmpty, "skip-empty", false, "Output nothing when there are no net-new findings")
	cmd.Flags().StringVar(&output, "output", "", "Write to this path (default: stdout)")
	return cmd
}

type prCommentEntry struct {
	path    string
	finding *core.Finding
}

func renderPRComment(runID string, rows []prCommentEntry) string {
	var b strings.Builder
	if len(rows) == 0 {
		fmt.Fprintf(&b, "## deepsec — no net-new findings for run `%s` :white_check_mark:\n", runID)
		return b.String()
	}
	by := map[core.Severity]int{}
	for _, r := range rows {
		by[r.finding.Severity]++
	}
	plural := ""
	if len(rows) != 1 {
		plural = "s"
	}
	fmt.Fprintf(&b, "## deepsec — %d net-new finding%s for run `%s`\n", len(rows), plural, runID)
	parts := []string{}
	for _, sev := range []core.Severity{core.SeverityCritical, core.SeverityHigh, core.SeverityHighBug, core.SeverityMedium, core.SeverityBug, core.SeverityLow} {
		if n, ok := by[sev]; ok {
			parts = append(parts, fmt.Sprintf("%d × %s", n, sev))
		}
	}
	fmt.Fprintf(&b, "%s\n\n", strings.Join(parts, " · "))
	for _, r := range rows {
		f := r.finding
		fmt.Fprintf(&b, "### `%s` — %s **%s**\n", sevBadge(f.Severity), f.Severity, f.Title)
		fmt.Fprintf(&b, "- file: `%s` (lines %v)\n", r.path, f.LineNumbers)
		fmt.Fprintf(&b, "- slug: `%s`\n", f.VulnSlug)
		fmt.Fprintf(&b, "- confidence: `%s`\n", f.Confidence)
		if f.Revalidation != nil {
			fmt.Fprintf(&b, "- revalidation: `%s`\n", f.Revalidation.Verdict)
		}
		fmt.Fprintf(&b, "\n%s\n\n", f.Description)
		fmt.Fprintf(&b, "**Fix:** %s\n\n", f.Recommendation)
	}
	return b.String()
}

func sevBadge(s core.Severity) string {
	switch s {
	case core.SeverityCritical:
		return "CRIT"
	case core.SeverityHigh, core.SeverityHighBug:
		return "HIGH"
	case core.SeverityMedium:
		return "MED "
	case core.SeverityBug:
		return "BUG "
	}
	return "LOW "
}
